### ISSUE TITLE
Intellisense completion

### ISSUE THREAD

**Issue by gabebw** [gabebw/dotfiles#163]
The options seem to be CoC, Deoplete, or YouCompleteMe.

My goals for Typescript:

* Show popup window with type information for arguments for a function, like `user.get(<TAB>` shows what arguments `user.get` expects
* Smart renaming of a given variable
* Autocomplete properties and method names based on type information
* Ability to turn off popup window (while keeping on-demand autocomplete) in Markdown

## CoC

https://github.com/neoclide/coc.nvim

✅ Works well for Typescript and autocompletes properties, module names in imports, etc
✅ Beautiful colored popup with autocomplete information
⚠️ Popup window can only be globally disabled/enabled, not per-filetype (I think). The setting is `"suggest.floatEnable": false` in `coc-settings.json` ([docs here](https://github.com/neoclide/coc.nvim/wiki/F.A.Q#how-could-i-disable-floating-window))...but that doesn't work, I still see the floating window in:

```
$ cat ~/.vim/coc-settings.json
{ "suggest.floatEnable": false }
```

Rust screenshot:
<img width="1361" alt="Alacritty 2020-05-17 21-14-53" src="https://user-images.githubusercontent.com/257678/82173827-7b00a900-9883-11ea-9294-48af697b3bf6.png">

## Deoplete

https://github.com/Shougo/deoplete.nvim

This GIF of its TypeScript functionality is nice: https://github.com/mhartington/nvim-typescript/raw/master/deoplete-tss.gif

It is very Python-heavy, and requires 2 extra plugins in Vim (but none in nvim):

```vim
Plug 'Shougo/deoplete.nvim'
Plug 'roxma/vim-hug-neovim-rpc'
Plug 'roxma/nvim-yarp'
```

Then it requires installing `pynvim` using the same version of Python that Vim is using. (This might not be required for neovim?)

Homebrew-installed Vim uses python@3.8, which is not linked by default, so to get it all working:

```
$ brew uninstall python
$ brew link --force python@3.8
$ pip3 install pynvim
```

I will say that after all that, it uses <kbd>Tab</kbd> very nicely with no additional settings.

✅ Works great with Rust and Typescript
❌ Preview window is not as pretty as CoC
❌ Requires a lot of Python code and setup that feels unrelated to Vim
⚠️  Deoplete also [enables/disables using autocommands](https://github.com/Shougo/deoplete.nvim/issues/464):

```vim
autocmd FileType markdown,text call deoplete#custom#buffer_option('auto_complete', v:false)
```

This makes it so the popup window doesn't appear on every word, but is triggered on `<Tab>`, which is fine.

## YouCompleteMe

https://github.com/ycm-core/YouCompleteMe

Haven't tried it yet.

To install with vim-plug: https://github.com/ycm-core/YouCompleteMe/issues/1751

**Comment by gabebw** [comment]
Hmmm, ALE completion actually gets me most of the way there. The popup is not nearly as beautiful as CoC's, but it's useful and requires almost no setup:

```vim
let g:ale_completion_enabled = 1
```

**Comment by gabebw** [comment]
Perhaps this will help: https://github.com/WilHall/.dotfiles/blob/main/.vim/rcplugins/coc.vim

**Comment by WilHall** [comment]
I also wrote a blog post on my setup: https://thoughtbot.com/blog/modern-typescript-and-react-development-in-vim

**Comment by gabebw** [comment]
@WilHall This is so helpful, thank you! I was lost in a ton of random CoC settings, but your blog post really clarified the minimum to have a good experience.

**Comment by gabebw** [comment]
Fixed in #185. Thank you @WilHall!
